pkg/webui/components/checkbox/group/index.js

```diff
@@ -0,0 +1,335 @@
+import Input from '../../input'
+import PropTypes from '../../../lib/prop-types'
+import debounce from '../../../lib/debounce'
+import FieldGroup, { formatFieldName } from '../../field/group'
+
+import style from './group.styl'
+
+const m = defineMessages({
+  checkAll: 'Check All',
+  selections: 'Your Selections',
+})
+
+@bind
+class CheckboxGroup extends React.Component {
+
+  constructor (props) {
+    super(props)
+
+    this.state = {
+      selected: [],
+      searchQuery: '',
+      allSelected: false,
+    }
+
+    if (props.searchDebounce) {
+      const { debouncedFunction, cancel } = debounce(
+        this.onSearchChange,
+        props.searchDebounce
+      )
+
+      this.debouncedSearch = debouncedFunction
+      this.debounceCancel = cancel
+    } else {
+      this.debouncedSearch = this.onSearchChange
+      this.debounceCancel = () => null
+    }
+  }
+
+  componentWillUnmount () {
+    this.debounceCancel()
+  }
+
+  onSelectedChange (name, value) {
+    const { onChange, children } = this.props
+    const totalCount = React.Children.count(children)
+
+    if (value) {
+      this.setState(function (prev) {
+        const selectedCount = prev.selected.length + 1
+
+        return {
+          selected: [ name, ...prev.selected ],
+          allSelected: totalCount === selectedCount,
+        }
+      })
+    } else {
+      this.setState(prev => ({
+        selected: prev.selected.filter(n => n !== name),
+        allSelected: false,
+      }))
+    }
+
+    if (onChange) {
+      onChange(name, value)
+    }
+  }
+
+  onTransferSelectedChange (name) {
+    const { onChange } = this.props
+
+    this.setState(prev => ({
+      selected: prev.selected.filter(n => n !== name),
+      allSelected: false,
+    }), function () {
+      if (onChange) {
+        onChange(name, false)
+      }
+    })
+  }
+
+  onSearchChange (value) {
+    this.setState({ searchQuery: value })
+  }
+
+  onSelectAll (_, value) {
+    const { setValues, children, name: groupName } = this.props
+    const { selected } = this.state
+
+    const fields = React.Children.toArray(children)
+    const newValues = {}
+    const newSelected = value ? [ ...selected ] : []
+
+    for (let i = 0; i < fields.length; i++) {
+      const name = fields[i].props.name
+      const fullName = formatFieldName(groupName, name)
+      newValues[name] = value
+      if (value && !selected.includes(fullName)) {
+        newSelected.push(fullName)
+      }
+    }
+
+    this.setState({
+      selected: newSelected,
+      allSelected: value,
+    }, () => setValues({ [groupName]: newValues }))
+  }
+
+  get header () {
+    const { transfer, selectAllTitle, selectAllName, children } = this.props
+    const { selected, allSelected } = this.state
+
+    const cls = classnames(style.header, {
+      [style.headerWithTransfer]: transfer,
+    })
+
+    const selectedCount = selected.length
+    const totalCount = React.Children.count(children)
```

Comment:
Might make sense to define this once in the beginning and then referencing. It shouldn't be changing, I guess?